pkg/peer/remote_track.go

```diff
@@ -0,0 +1,92 @@
+package peer
+
+import (
+	"errors"
+	"io"
+
+	"github.com/matrix-org/waterfall/pkg/common"
+	"github.com/pion/webrtc/v3"
+)
+
+func (p *Peer[ID]) handleNewVideoTrack(
+	trackInfo common.TrackInfo,
+	remoteTrack *webrtc.TrackRemote,
+	receiver *webrtc.RTPReceiver,
+) {
+	simulcast := common.RIDToSimulcastLayer(remoteTrack.RID())
+
+	// Notify others that our track has just been published.
+	p.state.AddRemoteTrack(remoteTrack)
+	p.sink.Send(NewTrackPublished{trackInfo, simulcast, nil})
+
+	// Start forwarding the data from the remote track to the local track,
+	// so that everyone who is subscribed to this track will receive the data.
+	go func() {
+		for {
+			packet, _, readErr := remoteTrack.ReadRTP()
+			if readErr != nil {
+				if readErr == io.EOF { // finished, no more data, no error, inform others
+					p.logger.Info("remote track closed")
+				} else { // finished, no more data, but with error, inform others
+					p.logger.WithError(readErr).Error("failed to read from remote track")
+				}
+				p.state.RemoveRemoteTrack(remoteTrack)
+				p.sink.Send(PublishedTrackFailed{trackInfo, simulcast})
+				return
+			}
+
+			p.sink.Send(RTPPacketReceived{trackInfo, simulcast, packet})
+		}
+	}()
+}
+
+func (p *Peer[ID]) handleNewAudioTrack(
+	trackInfo common.TrackInfo,
+	remoteTrack *webrtc.TrackRemote,
+	receiver *webrtc.RTPReceiver,
+) {
+	// Create a local track, all our SFU clients that are subscribed to this
+	// peer (publisher) wil be fed via this track.
+	localTrack, err := webrtc.NewTrackLocalStaticRTP(
+		remoteTrack.Codec().RTPCodecCapability,
+		remoteTrack.ID(),
+		remoteTrack.StreamID(),
+	)
+	if err != nil {
+		p.logger.WithError(err).Error("failed to create local track")
+		return
+	}
+
+	p.state.AddRemoteTrack(remoteTrack)
+	p.sink.Send(NewTrackPublished{trackInfo, common.SimulcastLayerNone, localTrack})
+
+	// Start forwarding the data from the remote track to the local track,
+	// so that everyone who is subscribed to this track will receive the data.
+	go func() {
+		rtpBuf := make([]byte, 1400)
+
+		// Call this when this goroutine ends.
+		defer func() {
+			p.sink.Send(PublishedTrackFailed{trackInfo, common.SimulcastLayerNone})
+			p.state.RemoveRemoteTrack(remoteTrack)
+		}()
+
+		for {
+			index, _, readErr := remoteTrack.Read(rtpBuf)
```

Comment:
Why not use `ReadRTP()`? It always felt a bit simpler to read to me

Reply:
I agree that `ReadRTP()` is a better choice when we're interested in RTP packets, that's also something that I use in when handling **video** tracks since we are working on a packet level when forwarding and rewriting packet IDs for the video.

In general, I think the only advantage of the `Read()` and `Write() over the `ReadRTP()` is that they use the buffer that you pass to the function, which means that `Read()` does not allocate, since we're reusing the same slice all over again (slight performance advantage). Since we don't really process audio packets in any way and just forward it within a single function, I just decided to use `Read()` / `Write()` here.

However, if you think that it harms the readability of this code, I can change it to the `ReadRTP()` and `WriteRTP()` (I think the performance loss, in this case, might be negligible anyway).